web3/utils/contracts.py

```diff
@@ -90,7 +90,10 @@ def find_matching_fn_abi(abi, fn_name=None, args=None, kwargs=None):
     if len(function_candidates) == 1:
         return function_candidates[0]
     if not function_candidates:
-        raise ValueError("No matching functions found")
+        raise ValueError(
+            "Incorrect number of arguments supplied or "
+            "one or more arguments could not be encoded to the necessary ABI type"
```

Comment:
Still leaves out the possibility that there are no functions by the given name.

- Simple solution: Add more *or* statements to this to add the extra wording.
- Change this logic to narrow down a matrix of what did/did-not match and tailor the error message to exactly what parts had no match.  Logic should not be to difficult to write, but it may take some churn to make it clean and easy to follow.